notebooks/working/investigating_deconfounder.py

```diff
@@ -0,0 +1,592 @@
+import matplotlib.pyplot as plt
+
+import seaborn as sns
+color_names = ["windows blue",
+               "amber",
+               "crimson",
+               "faded green",
+               "dusty purple",
+               "greyish"]
+colors = sns.xkcd_palette(color_names)
+sns.set(style="white", palette=sns.xkcd_palette(color_names), color_codes = False)
+
+
+import os
+os.listdir('.')
+
+# +
+drive_alone_graph = CausalGraphicalModel(
+    nodes=['a', 'b', 'c', 'confounder',
+          'd',  'e', 'f', 'y'],
+    edges=[
+        ("a", "y"), 
+        ("b", "y"), 
+        ("c", "y"), 
+        ("d", "y"), 
+        ("e", "y"), 
+        ("f", "y"), 
+        ("confounder", "y"), 
+        
+        
+        ("confounder", "a"),
+        ("confounder", "b"),
+#         ("confounder", "c"),
+        ("confounder", "d"),
+        ("confounder", "e"),
+        ("confounder", "f"),
+
+
+        
+        
+        ("confounder", "y"), 
+        ("confounder", "y"), 
+    ]
+)
+
+# draw return a graphviz `dot` object, which jupyter can render
+drive_alone_graph.draw()
+
+# +
+sample_size = 10000
+
+confounder = np.random.normal(loc=20, scale = 5, size = sample_size)
+
+# +
+a = .3*confounder + np.random.normal(loc=10,scale=3,size=sample_size)
+b = -1*confounder + np.random.normal(loc=6,scale=3,size=sample_size)
+c = np.random.normal(loc=15,scale=3,size=sample_size) # + .5*confounder 
+d = .8*confounder + np.random.normal(loc=-10,scale=3,size=sample_size)
+e = .5*confounder + np.random.normal(loc=8,scale=3,size=sample_size)
+f= -.3*confounder + np.random.normal(loc=-12,scale=3,size=sample_size)
+y = 5 + 2*a + 2*b - 5*c + 3*d + 2*e - 5*f + 7*confounder + np.random.normal(loc=0,scale=1,size=sample_size)
+
+coeffs_true = [5,2,2,-5,3,2,-5,7]
+
+# plt.hist(confounder, bins = 50)
+
+# plt.hist(f, bins = 50)
+# -
+
+df = pd.DataFrame(np.array([a,b,c,d,e,f,confounder, y]).T, columns = ['a', 'b', 'c',
+          'd',  'e', 'f', 'confounder', 'y'])
+df.head()
+
+# +
+# sns.pairplot(df, size=1.5)
+
+# +
+ols_formula = "y ~ a+b+c+d+e+f"#+confounder "
+
+model_full = smf.ols(ols_formula, data=df)
+results_full = model_full.fit()
+results_df = pd.DataFrame(results_full.params, columns = ['estimated_params'])
+results_df['std_errors'] = results_full.bse
+results_df['true_params'] = coeffs_true[:-1]
+results_df
+# -
+
+model_partial = smf.ols("y ~ a+b+c+d+e+f ", data=df)
+results_partial = model_partial.fit()
+results_df_partial = pd.DataFrame(results_partial.params, columns = ['estimated_params'])
+results_df_partial['std_errors'] = results_partial.bse
+results_df_partial['true_params'] = coeffs_true[:-1]
+results_df_partial
+
+# ## Factor analysis: Is there evidence of latent factors?
+
+from factor_analyzer.factor_analyzer import calculate_bartlett_sphericity,calculate_kmo
+
+chi_square_value,p_value=calculate_bartlett_sphericity(df[['a','b','c','d','e','f']])
+print("Bartlett sphericity test indicates potential latent confounder, with a p-value of: ", p_value)
```

Comment:
According to wikipedia, the Bartlett test is testing for the assumption of equal variances amongst the k-column vectors. If one standardizes the variables inside the function before testing, then you can say the Bartlett test is testing for the assumption of k-independent, normally distributed vectors.

We should probably be specific since departures from normality may cause this test to fail, even if one has no unobserved latent confounders.